### PR TITLE
AbstractOperators for agentic patterns

### PR DESCRIPTION
**Pull request by mahikaw**
## Description                   
This PR introduces three new agentic retrieval operators built on top of `AbstractOperator`, plus an end-to-end example pipeline:

- **SubQueryGeneratorOperator** — expands each query row into multiple sub-query rows via an LLM. Supports three built-in strategies: `decompose` (split into complementary sub-aspects), `hyde` (generate hypothetical answer passages), and
    `multi_perspective` (rephrase from diverse angles). Downstream retrieval operators retrieve independently per sub-query, enabling RRF-style aggregation.

- **ReActAgentOperator** — implements an agentic ReAct (Reason + Act) retrieval loop. The LLM iteratively reasons about the query and issues targeted retrieval calls, accumulating documents across steps until it decides it has enough
    evidence or hits a step limit. Supports parallel query execution and configurable step budgets.

- **SelectionAgentOperator** — re-ranks a set of retrieved candidate documents using an agentic LLM loop. The LLM is given the query and candidate docs, then calls a `log_selected_documents` tool to produce a final ranked list. Supports a
    `think` tool for extended reasoning.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/nv-ingest/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] If adjusting docker-compose.yaml environment variables have you ensured those are mimicked in the Helm values.yaml file.

### CODE REVIEW

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+    Output DataFrame schema
+    -----------------------
+    query_id : str  — same ``query_id`` as the input
+    doc_id   : str  — selected document ID
+    rank     : int  — 1-indexed rank (1 = most relevant)
+    message  : str  — LLM explanation of the selection
+
+    Parameters
+    ----------
+    llm_model : str
```
Comment (jperez999) [inline nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py:145]: so this will follow openai api?

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+            raise ImportError(
+                "SelectionAgentOperator requires 'openai'. " "Install it with:  pip install 'openai>=1.0'"
+            ) from exc
+
+        api_key = self._api_key
+        if api_key is not None and api_key.strip().startswith("os.environ/"):
+            var = api_key.strip().removeprefix("os.environ/")
+            api_key = os.environ[var]
+
+        self._client = OpenAI(
```
Comment (jperez999) [inline nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: So do we need to do this with openai? we have some nim communicators that follow the same openai api but they remove the openai package requirement which was an ask from product a while back. That is why you had to add it back. Check out https://github.com/NVIDIA/NeMo-Retriever/blob/1ae0ed753edde8c4df9a196902689d50852c684f/nemo_retriever/src/nemo_retriever/nim/nim.py if it doesnt have what you need, we can talk further.

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+            tool_choice="auto",
+        )
+        if self._max_tokens is not None:
+            call_kwargs["max_tokens"] = self._max_tokens
+
+        for _step in range(self._max_steps):
+            response = self._client.chat.completions.create(messages=messages, **call_kwargs)
+            choice = response.choices[0]
+            msg = choice.message
+
```
[the remark this answers is not in the document]
Comment (jperez999) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: This might be something to check.

**Review by jperez999 — CHANGES_REQUESTED** [review]
Great first run. Lets add some unit tests for expected data the might come in (and error edge cases) for each of the operators.

**Comment by mahikaw** [comment]
@greptileai this pr should aim to use NIM instead of openAI APIs. with that updated understanding, re-review it.

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: addressed

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+        api_key = self._api_key
+        if api_key is not None and api_key.strip().startswith("os.environ/"):
+            var = api_key.strip().removeprefix("os.environ/")
+            api_key = os.environ[var]
+
+        self._client = OpenAI(
+            api_key=api_key,
+            **({"base_url": self._base_url} if self._base_url is not None else {}),
+        )
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: addressed

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            from openai import OpenAI
+        except ImportError as exc:
+            raise ImportError(
+                "SelectionAgentOperator requires 'openai'. " "Install it with:  pip install 'openai>=1.0'"
+            ) from exc
+
+        api_key = self._api_key
+        if api_key is not None and api_key.strip().startswith("os.environ/"):
+            var = api_key.strip().removeprefix("os.environ/")
+            api_key = os.environ[var]
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: addressed

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+                elif tc.function.name == "log_selected_documents":
+                    doc_ids: List[str] = fn_args.get("doc_ids", [])
+                    # Filter to only valid candidates
+                    doc_ids = [d for d in doc_ids if d in valid_id_set][:feasible_k]
+                    end_kwargs = {
+                        "doc_ids": doc_ids,
+                        "message": fn_args.get("message", ""),
+                    }
+                    should_end = True
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py]: added warning with count, query text, and bad doc IDs

File: nemo_retriever/src/nemo_retriever/graph/subquery_operator.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    text = text.strip()
+
+    try:
+        parsed = json.loads(text)
+        if isinstance(parsed, list) and parsed and all(isinstance(s, str) for s in parsed):
+            return parsed
+    except json.JSONDecodeError:
+        pass
+
+    return [fallback]
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/subquery_operator.py:364]: fixed

File: nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-25, NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/selection_agent_operator.py:3]: added more coverage in preprocess steps and other helper functions

File: nemo_retriever/src/nemo_retriever/graph/react_agent_operator.py [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+            if len(results) >= self._retriever_top_k:
+                break
+
+        return results
+
+    def _resolve_api_key(self) -> Optional[str]:
+        api_key = self._api_key
+        if api_key is not None and api_key.strip().startswith("os.environ/"):
+            var = api_key.strip().removeprefix("os.environ/")
+            return os.environ[var]
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/react_agent_operator.py]: fixed earlier

File: nemo_retriever/src/nemo_retriever/graph/subquery_operator.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+            messages=[
+                {"role": "system", "content": system_prompt},
+                {"role": "user", "content": f"Query: {query}"},
+            ],
+        )
+        if self._max_tokens is not None:
+            call_kwargs["max_tokens"] = self._max_tokens
+
+        response = self._client.chat.completions.create(**call_kwargs)
+        raw = response.choices[0].message.content.strip()
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/subquery_operator.py]: fixed

File: nemo_retriever/src/nemo_retriever/graph/react_agent_operator.py [hunk 11]
```diff
@@ -0,0 +1,679 @@
...
+            with ThreadPoolExecutor(max_workers=min(self._num_concurrent, len(query_rows))) as executor:
+                futures = {
+                    executor.submit(self._run_single_query, qid, qtxt, api_key): (qid, qtxt) for qid, qtxt in query_rows
+                }
+                for future in as_completed(futures):
+                    try:
+                        rows.extend(future.result())
+                    except Exception as exc:
+                        qid, qtxt = futures[future]
+                        logger.warning("ReActAgentOperator: query %r failed: %s", qid, exc)
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/src/nemo_retriever/graph/react_agent_operator.py]: fixed

File: nemo_retriever/tests/test_agentic_operators.py [hunk 12]
```diff
@@ -0,0 +1,380 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-25, NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Smoke tests for the agentic retrieval operators.
+
+Run with:
+    cd nemo_retriever && uv run pytest tests/test_agentic_operators.py -v
+"""
+
```
[the remark this answers is not in the document]
Comment (mahikaw) [reply nemo_retriever/tests/test_agentic_operators.py:10]: added 2 and 3, other already fixed.

File: nemo_retriever/src/nemo_retriever/graph/subquery_operator.py [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+                messages_list=[messages],
+                model=self._llm_model,
+                api_key=self._resolve_api_key(),
+                extra_body=extra_body or None,
+            )
+        except Exception as exc:
+            logger.warning("SubQueryGeneratorOperator: LLM call failed for query %r: %s", query, exc)
+            return [query]
+        raw = results[0].strip()
+        return _parse_json_list(raw, fallback=query)
```
[the remark this answers is not in the document]
Comment (jperez999) [reply nemo_retriever/src/nemo_retriever/graph/subquery_operator.py:332]: Please adjust this exception to make it more robust.

**Review by jperez999 — CHANGES_REQUESTED** [review]

**Review by jperez999 — APPROVED** [review]
